src/main/scala/com/fulcrumgenomics/umi/GroupReadsByUmi.scala

```diff
@@ -554,31 +562,39 @@ class GroupReadsByUmi
   }
 
   /**
-    * Retrieves the UMI to use for a read pair.  In the case of single-umi strategies this is just
+    * Retrieves the UMI to use for a template.  In the case of single-umi strategies this is just
     * the upper-case UMI sequence.
     *
     * For Paired the UMI is extended to encode which "side" of the template the read came from - the earlier
     * or later read on the genome.  This is necessary to ensure that, when the two paired UMIs are the same
     * or highly similar, that the A vs. B groups are constructed correctly.
     */
-  private def umiForRead(r1: SamRecord, r2: SamRecord): Umi = r1.get[String](this.rawTag) match {
-    case None | Some("") => fail(s"Record '$r1' was missing the raw UMI tag '${this.rawTag}'")
-    case Some(chs)       =>
-      val umi = chs.toUpperCase
+  private def umiForRead(t: Template): Umi = {
+    // Check that all the primary reads have the UMI defined
+    t.primaryReads.foreach { rec =>
+      val umi = rec.get[String](this.rawTag)
+      if (!umi.exists(_.nonEmpty)) fail(s"Record '$rec' was missing the raw UMI tag '${this.rawTag}'")
+    }
 
-      this.assigner match {
-        case paired: PairedUmiAssigner =>
-          require(r1.refIndex == r2.refIndex, s"Mates on different references not supported: ${r1.name}")
-          val pos1 = if (r1.positiveStrand) r1.unclippedStart else r1.unclippedEnd
-          val pos2 = if (r2.positiveStrand) r2.unclippedStart else r2.unclippedEnd
-          val r1Lower = pos1 < pos2 || pos1 == pos2 && r1.positiveStrand
-          val umis = umi.split('-')
-          require(umis.length == 2, s"Paired strategy used but umi did not contain 2 segments: $umi")
-
-          if (r1Lower) paired.lowerReadUmiPrefix  + ":" + umis(0) + "-" + paired.higherReadUmiPrefix + ":" + umis(1)
-          else         paired.higherReadUmiPrefix + ":" + umis(0) + "-" + paired.lowerReadUmiPrefix  + ":" + umis(1)
-        case _ =>
-          umi
-      }
+    if (t.r1.isEmpty) fail(s"Read 1 missing for template ${t.name}.")
+
+    val umi = t.r1.get[String](this.rawTag)
```

Comment:
what's the type here?  I thought `Template.r1` was an `Option[SamRecord]`

Reply:
Is there an implicit here or something?

Reply:
This is one of the few cases where I wish there were a better way to differentiate what's happening. This is the equivalent of:

```scala
val r1 = t.r1.get
val umi = r1[String](this.rawTag)
```

I'll change the `.get` to `.getOrElse(unreachable("R1 must be present"))` which should make it a little clearer